zio-http/src/main/scala/zio/http/ServerConfig.scala

```diff
@@ -0,0 +1,182 @@
+package zio.http
+
+import io.netty.bootstrap.ServerBootstrap
+import io.netty.channel.ChannelPipeline
+import io.netty.util.ResourceLeakDetector
+import zio.http.service.{EventLoopGroup, ServerChannelFactory}
+import zio.{ULayer, ZIO, ZLayer}
+import zio.http.service.ServerSSLHandler.ServerSSLOptions
+
+import java.net.{InetAddress, InetSocketAddress}
+
+object ServerConfig {
+
+  val default: ZLayer[Any, Nothing, Config with EventLoopGroup with ServerChannelFactory] = {
+    val configLayer: ULayer[Config] = ZLayer.succeed(Config())
+    configLayer ++ EventLoopGroup.auto(0) ++ ServerChannelFactory.auto
+  }
+
+  def live(config: Config): ZLayer[Any, Any, Config with EventLoopGroup with ServerChannelFactory] = {
+    val (eventLoopGroupLayer, serverChannelFactoryLayer) = config.channelType match {
+      case ChannelType.NIO => (EventLoopGroup.nio(config.nThreads), ServerChannelFactory.nio)
+      case ChannelType.EPOLL => (EventLoopGroup.epoll(config.nThreads), ServerChannelFactory.epoll)
+      case ChannelType.KQUEUE => (EventLoopGroup.kQueue(config.nThreads), ServerChannelFactory.kQueue)
+      case ChannelType.URING => (EventLoopGroup.uring(config.nThreads), ServerChannelFactory.uring)
+      case ChannelType.AUTO => (EventLoopGroup.auto(config.nThreads), ServerChannelFactory.auto)
+    }
+    val configLayer: ULayer[Config] = ZLayer.succeed(config)
+    configLayer ++ eventLoopGroupLayer ++ serverChannelFactoryLayer
+  }
+
+  val testConfig: ZLayer[Any, Nothing, Config with EventLoopGroup with ServerChannelFactory] = {
+    val configLayer: ULayer[Config] = ZLayer.succeed(Config().withPort(0).withLeakDetection(LeakDetectionLevel.PARANOID))
+    configLayer ++ EventLoopGroup.nio(0) ++ ServerChannelFactory.nio
+  }
+
+  sealed trait ChannelType
+
+  object ChannelType {
+    case object NIO extends ChannelType
+
+    case object EPOLL extends ChannelType
+
+    case object KQUEUE extends ChannelType
+
+    case object URING extends ChannelType
+
+    case object AUTO extends ChannelType
+  }
+
+  sealed trait LeakDetectionLevel {
+    self =>
+    def jResourceLeakDetectionLevel: ResourceLeakDetector.Level = self match {
+      case LeakDetectionLevel.DISABLED => ResourceLeakDetector.Level.DISABLED
+      case LeakDetectionLevel.SIMPLE => ResourceLeakDetector.Level.SIMPLE
+      case LeakDetectionLevel.ADVANCED => ResourceLeakDetector.Level.ADVANCED
+      case LeakDetectionLevel.PARANOID => ResourceLeakDetector.Level.PARANOID
+    }
+  }
+
+  object LeakDetectionLevel {
+    case object DISABLED extends LeakDetectionLevel
+
+    case object SIMPLE extends LeakDetectionLevel
+
+    case object ADVANCED extends LeakDetectionLevel
+
+    case object PARANOID extends LeakDetectionLevel
+  }
+
+  final case class Config(
+                           leakDetectionLevel: LeakDetectionLevel = LeakDetectionLevel.SIMPLE,
+                           error: Option[Throwable => ZIO[Any, Nothing, Unit]] = None,
```

Comment:
I think we should move this to `Server`, something like, `Server#onError`. Because it's a function, not data, and unlike the rest of the config, it cannot be read from external configuration.

Reply:
@jdegoes I have pushed the change but implemented differently from your suggestion. OnError callback is more like an input parameter for the Server layer.